doc/api/stream.md

```diff
@@ -1112,22 +1290,112 @@ class SimpleProtocol extends Readable {
 // with the parsed header data.
 ```
 
-### Class: stream.Transform
+### Implementing a Duplex Stream
+
+A [Duplex][] stream is one that implements both [Readable][] and [Writable][],
+such as a TCP socket connection.  
+
+Because Javascript does not have support for multiple inheritance, the 
+`stream.Duplex` class is extended to implement a [Duplex][] stream (as opposed
+to extending the `stream.Readable` *and* `stream.Writable` classes).
+
+*Note*: The `stream.Duplex` class prototypically inherits from `stream.Readable`
+and parasitically from `stream.Writable`.
+
+Custom Duplex streams *must* call the `new stream.Duplex([options])`
+constructor and implement *both* the `readable.\_read()` and
+`writable.\_write()` methods.
+
+#### new stream.Duplex(options)
+
+* `options` {Object} Passed to both Writable and Readable
+  constructors. Also has the following fields:
+  * `allowHalfOpen` {Boolean} Default = `true`. If set to `false`, then
+    the stream will automatically end the readable side when the
+    writable side ends and vice versa.
+  * `readableObjectMode` {Boolean} Default = `false`. Sets `objectMode`
+    for readable side of the stream. Has no effect if `objectMode`
+    is `true`.
+  * `writableObjectMode` {Boolean} Default = `false`. Sets `objectMode`
+    for writable side of the stream. Has no effect if `objectMode`
+    is `true`.
+
+For example:
+
+```js
+const Duplex = require('stream').Duplex;
+
+class MyDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+  }
+}
+```
+
+Or, when using pre-ES6 style constructors:
+
+```js
+const Duplex = require('stream').Duplex;
+const util = require('util');
+
+function MyDuplex(options) {
+  if (!(this instanceof MyDuplex))
+    return new MyDuplex(options);
+  Duplex.call(this, options);
+}
+util.inherits(MyDuplex, Duplex);
+```
+
+#### An Example Duplex Stream
+
+The following illustrates a simple example of a Duplex stream that buffers
+incoming written data via the [Writable][] interface that is read back out
+via the [Readable][] interface.
+
+```js
+const Duplex = require('stream').Duplex;
+const kBuf = Symbol('buffer');
+
+class MyDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+  }
+
+  _write(chunk, encoding, next) {
+    var buffer = this[kBuf];
+    if (!Array.isArray(buffer))
+      buffer = this[kBuf] = [];
+    buffer.push(chunk);
+    next();
+  }
+
+  _read(size) {
+    var buffer = this[kBuf];
+    if (Array.isArray(buffer))
+      buffer.forEach((chunk) => this.push(chunk));
+  }
+}
+```
```

Comment:
This might be buffer overkill. `Duplex` instances already come with two buffers: one for the writable side and one for the readable side. Since the above example operations synchronously, this never comes into play ... _however_, were the operations to become asynchronous, all three buffers would be involved!

It is pretty difficult to find a good use for `Duplex` that isn't fulfilled by `Transform`, or already implemented by core — generally there's a low-level resource that represents a `fd` with disconnected input and output, that might be okay to have half-open. I'd be tempted to say that we should create a [duplexer](http://npm.im/duplexer)-type example here, where a pipeline of streams is represented to the outside world as a single stream, which is a _super_ handy pattern in practice. However, I almost think we'd be better off using `stream.Transform` to illustrate that.